tests/Unit/EmogrifierTest.php

```diff
@@ -737,6 +737,117 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+            ':not with type more specific than nothing' => ['<p class="p-1"', '.p-1', '.p-1:not(h1)'],
+            ':not with several types more specific than single type' => ['<p class="p-1"', 'p', '*:not(html body h1)'],
+            ':not with class more specific than types' => ['<p class="p-1"', 'html body p', '*:not(.p-2)'],
+            ':not with class more specific than :not with types' => [
+                '<p class="p-1"',
+                '*:not(html body h1)',
+                '*:not(.p-2)',
+            ],
+            'class more specific than :not with types' => ['<p class="p-1"', '*:not(html body h1)', '.p-1'],
+        ];
+    }
+
+    /**
+     * @test
+     * @param string $matchedTagPart Tag expected to be matched by both selectors, without the closing '>',
+     *                               e.g. '<p class="p-1"'
+     * @param string $lessSpecificSelector A selector expression
+     * @param string $moreSpecificSelector Some other, more specific selector expression
+     * @dataProvider differentCssSelectorSpecificityDataProvider
+     */
+    public function emogrifyAppliesMoreSpecificCssSelectorToMatchingElements(
+        $matchedTagPart,
+        $lessSpecificSelector,
+        $moreSpecificSelector
+    ) {
+        $this->subject->setHtml(static::COMMON_TEST_HTML);
+        $this->subject->setCss(
+            $lessSpecificSelector . ' { color: red; } ' .
+            $moreSpecificSelector . ' { color: green; } ' .
+            $moreSpecificSelector . ' { background-color: green; } ' .
+            $lessSpecificSelector . ' { background-color: red; }'
+        );
+
+        $result = $this->subject->emogrify();
+
+        static::assertContains($matchedTagPart . ' style="color: green; background-color: green;"', $result);
+    }
+
+    /**
+     * @test Confirms the dataset is valid for the test, and the less specific selector does in fact match
+     * @param string $matchedTagPart Tag expected to be matched by selector, without the closing '>',
+     *                               e.g. '<p class="p-1"'
+     * @param string $lessSpecificSelector A selector expression
+     * @dataProvider differentCssSelectorSpecificityDataProvider
+     */
+    public function emogrifyWouldApplyLessSpecificCssSelectorToMatchingElements(
+        $matchedTagPart,
+        $lessSpecificSelector
+    ) {
+        $this->subject->setHtml(static::COMMON_TEST_HTML);
+        $this->subject->setCss($lessSpecificSelector . ' { color: green; }');
+
+        $result = $this->subject->emogrify();
+
+        static::assertContains($matchedTagPart . ' style="color: green;"', $result);
+    }
+
+    /**
+     * @return string[][]
+     */
+    public function equalCssSelectorSpecificityDataProvider()
+    {
+        return [
+            'pseudo-class as specific as class' => ['<p class="p-1"', '.p-1', '*:first-child'],
+            'attribute as specific as class' => ['<span title="bonjour"', 'span[title="bonjour"]', '.p-2 span'],
```

Comment:
Should we also have a case comparing a pseudo-class comparing an attribute, or would that be redundant?